src/ui/textfield.rs

```diff
@@ -85,56 +143,191 @@ impl TextField {
             canvas.draw_rrect(rrect, &paint);
         });
 
-        // rendering: text
+        // Rendering: text
         ui.push_group(ui.size(), Layout::Freeform);
         ui.pad((16.0, 0.0));
         canvas.save();
         ui.clip(canvas);
 
-        // render hint
+        // Rendering: hint
         if hint.is_some() && self.text.len() == 0 {
             ui.text(canvas, hint.unwrap(), colors.text_hint, (AlignH::Left, AlignV::Middle));
         }
-        let text_advance = ui.text(canvas, &self.text_utf8, colors.text, (AlignH::Left, AlignV::Middle));
+
+        ui.text(canvas, &self.text_utf8, colors.text, (AlignH::Left, AlignV::Middle));
 
         if self.focused && (input.time_in_seconds() - self.blink_start) % Self::BLINK_PERIOD < Self::HALF_BLINK {
             ui.draw_on_canvas(canvas, |canvas| {
                 let mut paint = Paint::new(Color4f::from(colors.text), None);
                 paint.set_anti_alias(false);
                 paint.set_style(paint::Style::Stroke);
-                let x = text_advance + 1.0;
+
+                let current_text: String = self.text[..self.selection.first].iter().collect();
+                let current_text_width = ui.borrow_font().measure_str(current_text, None).0;
+
+                let x = current_text_width + 1.0;
                 let y1 = Self::height(ui) * 0.2;
                 let y2 = Self::height(ui) * 0.8;
                 canvas.draw_line((x, y1), (x, y2), &paint);
             });
         }
 
+        if self.selection.first != self.selection.second {
+            ui.draw_on_canvas(canvas, |canvas| {
+                let mut paint = Paint::new(Color4f::from(colors.selection), None);
+                paint.set_anti_alias(true);
+
+                // Get all text from textfield start to current selection cursor position.
+                // This will act as base position for selection.
+                let selection_anchor_text: String = self.text[..self.selection.start()].iter().collect();
+                let selection_anchor_text_width = ui.borrow_font().measure_str(selection_anchor_text, None).0;
+
+                // Get text left to right or right to left depending on end cursor position.
```

Comment:
This comment is a bit outdated and describes behavior from an older revision of this PR. Right now this code doesn't care about relative positioning of the cursor vs the anchor, as that's done by `Selection::normalize()`.